### PR TITLE
refactor: ts sdk module execution system to use imports

### PR DESCRIPTION
**Pull request by TomChv**
Rework the way the user's code is imported to be executed to actually uses `imports` instead of loading it from its decorators.

This would unblock severals features such as interface supports and compositions.

It's also a safer way to actually executes the user's code and resolve default value by reference, specially complex values since they can now be resolve by actually executing them if they are exported.

The changes are backward compatible.

### CODE REVIEW

**Comment by jedevc** [comment]
Is there a way we can make this change without breaking modules that have already been published to the daggerverse? Can we change the behavior based on the `engineVersion`?

**Comment by TomChv** [comment]
> Is there a way we can make this change without breaking modules that have already been published to the daggerverse? Can we change the behavior based on the engineVersion?

I could try to resolve the execution using the registry first, then the executor if it's failing to keep things backward compatible, but that's quite ugly no?

**Comment by TomChv** [comment]
@helderco This is ready for a review! Only the PHP tests are failing which isn't related to this PR

File: core/integration/module_typescript_test.go [hunk 1]
```diff
@@ -1304,3 +1305,30 @@ class Test {
...
+			With(daggerExec("init", "--name=test", "--sdk=typescript")).
+			With(sdkSource("typescript", `
+import { func, object } from "@dagger.io/dagger"
+
+@object()
+class Test {
+  @func()
+	foo(): string {
+		return "bar"
+	}
```
Comment (helderco) [inline core/integration/module_typescript_test.go:1325]:
```suggestion
  @func()
  foo(): string {
    return "bar"
  }
```

File: sdk/typescript/entrypoint/invoke.ts [hunk 2]
```diff
@@ -38,15 +44,34 @@ export async function invoke(module: DaggerModule, ctx: InvokeCtx) {
...
+    if (e instanceof FunctionNotFound) {
+      result = await registry.getResult(
+        object.name,
+        method.name,
+        parentState,
+        args,
+      )
+    } else {
+      throw e
+    }
```
[the remark this answers is not in the document]
Comment (helderco) [reply sdk/typescript/entrypoint/invoke.ts:73]: I see that you did add backwards compat.

**Review by helderco — COMMENTED** [review]
Waiting for reply on https://github.com/dagger/dagger/pull/8653#discussion_r1799701645

File: core/integration/module_type_test.go [hunk 3]
```diff
@@ -73,7 +73,7 @@ def repeater(msg: str, times: int) -> Repeater:
 import { object, func } from "@dagger.io/dagger"
 
 @object()
-class Repeater {
+export class Repeater {
```
[the remark this answers is not in the document]
Comment (helderco) [reply core/integration/module_type_test.go:76]: Don't merge before discussing this.

File: sdk/typescript/entrypoint/invoke.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (TomChv) [reply sdk/typescript/entrypoint/invoke.ts:73]: Yeah I'm gonna update it

File: core/integration/module_type_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (TomChv) [reply core/integration/module_type_test.go:76]: With the `import` logic, everything that should be accessible to the runtime must be exported, this exclude every existing classes in the module that should be exposed etc...

**Comment by TomChv** [comment]
> Waiting for reply on https://github.com/dagger/dagger/pull/8653#discussion_r1799701645

Replied :D

**Pull request merged**